Dockerfile

```diff
@@ -1,4 +1,4 @@
-# syntax=docker/dockerfile:1.1.7-experimental
+# syntax=docker/dockerfile:1.2
```

Comment:
@tonistiigi wondering; does BuildKit detect what version of the front-end it has "built-in"? In other words; if I'm building on docker 20.10, would it skip pulling the dockerfile frontend? What if I'm building on (e.g.) docker 21.xx, would it _downgrade_ the frontend to an older version?

If we want to consider adding a `# syntax` directive to be "best practice", we should probably describe the behaviour (and possible consequences for air-gapped setups)

Reply:
The image is always used when specified.

> and possible consequences for air-gapped setups

As it is an image defined in a Dockerfile it behaves the same as other images defined here. If you want to build in "air-gapped" setup you need to preload all images.

Reply:
The short answer is it will always look for `docker/dockerfile:1.2`. The frontend is a hidden image, so there's no downgrading, only pulling. If you're saying that the build-in frontend in 21.xx would have more features than 1.2, sure. But it's once people start needing the newer features that they may upgrade to 1.3, until then 1.2 is fine. They'll pick up bugfixes and security fixes.

I thought there was an issue for air-gapped setups but I can't find it. Either way, this conversation will be lost in this PR.

Reply:
> The image is always used when specified.

Ah, right; I was wondering if it would do a "match", and skip pulling if the syntax is the same as the default one

> so there's no downgrading, only pulling

Gotcha; makes sense. Mostly wondered (given that the version is SemVer(isn), if there was some logic that would skip pulling if it's already "ahead". Guess it makes sense to not do so, for (corner) cases where newer versions accidentally broke backward-compat.